### PR TITLE
Fw steps

### PR DESCRIPTION
**Pull request by ronhyman**
separate .ino file that detects good data, and has logic to send it to backend! next step is to implement api stuffs

### CODE REVIEW

File: data_analysis/data_capture.py [hunk 1]
```diff
@@ -88,5 +88,5 @@ def collect_data(seconds: float = None, serial_port = '/dev/cu.usbserial-0001',
     vibes, events = collect_data(int(seconds), fs=env.fs)
 
     rec = Recording(env, events, vibes)
-    rec.to_yaml(f'datasets/{timestamp}.yaml')
+    # rec.to_yaml(f'datasets/{timestamp}.yaml')
```
Comment (draymond63) [inline data_analysis/data_capture.py]: You should always look at what files you're commiting. You should not have commited this or the `normal_detection.html`

File: fw/bno055_steps/bno055_steps.ino [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+Adafruit_BNO055 bno = Adafruit_BNO055(55, 0x28);
+ESP8266Timer i_timer;  // Hardware Timer
+
+#define SAMPLE_RATE 100 // Sample rate for accelerometer in Hz
+#define INTERRUPT_INTERVAL_US (1000000/SAMPLE_RATE) // Interval between samples in microseconds
+#define START_BUFFER_TIME 1 // Size of buffer to store acceleration data prior to first step (in seconds)
+#define END_BUFFER_TIME 2 // Size of buffer to store acceleration data after last step (in seconds)
+#define START_BUFFER_SAMPLES START_BUFFER_TIME * SAMPLE_RATE  // In samples
+#define END_BUFFER_SAMPLES END_BUFFER_TIME * SAMPLE_RATE // Number of bad samples to save before stopping saving data
+#define AMP_THRESHOLD 0.5 // Threshold for amplitude of acceleration data to be considered "good" (in m/s^2)
```
Comment (draymond63) [inline fw/bno055_steps/bno055_steps.ino:26]: 😍

File: fw/bno055_steps/bno055_steps.ino [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+volatile bool start_sampling = false; // Flag to indicate if a new sample should be taken based on timer interrupt
+
+float sensor_data_buffer[START_BUFFER_SAMPLES]; // Circular buffer to store acceleration data
+int buffer_index = 0; // Index of circular buffer
+String post_data = ""; // String to store data to be sent to backend server
+bool save_sample = false; // Flag to indicate if a sample should be saved to post_data
+bool post_data_ready = false; // Flag to indicate if post_data is ready to be sent to backend server
+int bad_data_count = 0; // Counter to logic to stop saving data to post_data if too many bad samples are received
```
Comment (draymond63) [inline fw/bno055_steps/bno055_steps.ino]: I don't like this variable name. `end_buffer_length`?

File: fw/bno055_steps/bno055_steps.ino [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+#define START_BUFFER_TIME 1 // Size of buffer to store acceleration data prior to first step (in seconds)
+#define END_BUFFER_TIME 2 // Size of buffer to store acceleration data after last step (in seconds)
+#define START_BUFFER_SAMPLES START_BUFFER_TIME * SAMPLE_RATE  // In samples
+#define END_BUFFER_SAMPLES END_BUFFER_TIME * SAMPLE_RATE // Number of bad samples to save before stopping saving data
+#define AMP_THRESHOLD 0.5 // Threshold for amplitude of acceleration data to be considered "good" (in m/s^2)
+
+volatile bool start_sampling = false; // Flag to indicate if a new sample should be taken based on timer interrupt
+
+float sensor_data_buffer[START_BUFFER_SAMPLES]; // Circular buffer to store acceleration data
+int buffer_index = 0; // Index of circular buffer
```
Comment (draymond63) [inline fw/bno055_steps/bno055_steps.ino]: We've moved towards a "double buffer" way of thinking about it. Maybe this variable should be called `start_buffer_index`?

File: fw/bno055_steps/bno055_steps.ino [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    good_sample = abs(event.acceleration.z) > AMP_THRESHOLD;
+
+    // if sample is considered good (above threshold), save buffer to string and start saving data to post_data
+    if (good_sample && !save_sample) {
+      Serial.println("Begin saving data");
+      // save buffer to string
+      for (int i = 0; i < START_BUFFER_SAMPLES; i++) {
+          post_data += String(sensor_data_buffer[i]) + ",";
+      }
+      save_sample = true;
```
Comment (draymond63) [inline fw/bno055_steps/bno055_steps.ino]: As the code gets more complex, you should start migrating sections of it into functions (even if they are only used once). It help improve the readability. Right now, the `loop` function is very complicated.

**Review by draymond63 — CHANGES_REQUESTED** [review]
Remove the useless changes/files :)

File: data_analysis/data_capture.py (hunk 1 above)
Comment (ronhyman) [reply data_analysis/data_capture.py]: word - how do I undo those changes?

File: fw/bno055_steps/bno055_steps.ino (hunk 5 above)
Comment (ronhyman) [reply fw/bno055_steps/bno055_steps.ino]: good point I added a few functions to shorten loop()!

**Review by draymond63 — APPROVED** [review]
Lovely :)

**Pull request merged**